option/resolver: Avoid falling back to v2alpha1 CNC if v2 found

christarazi opened the pull request:
- **option/resolver: Use structured logging for errors**
- **option,resolver: Sort config sources descriptions**
- **option/resolver: Avoid falling back to v2alpha1 CNC if v2 found**

Fixes: https://github.com/cilium/cilium/pull/31721

christarazi committed:
option/resolver: Use structured logging for errors

Fixes: db0d41af34 ("crd, option: Make CiliumNodeConfig available on v2 API")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
option/resolver: Avoid falling back to v2alpha1 CNC if v2 found

Previously, Cilium would always fetch the v2alpha1 version of the CNC
even if the v2 version was already found. Fetching the v2alpha1 triggers
the deprecation warning from Kubernetes as it was deprecated in
https://github.com/cilium/cilium/pull/31721. Those warnings cause CI
failures such as:

```
FAIL: Found 26 k8s-app=cilium logs matching list of errors that must be investigated:
2024-06-17T13:38:56.846200595Z time="2024-06-17T13:38:56Z" level=warning msg="cilium.io/v2alpha1 CiliumNodeConfig will be deprecated in cilium v1.16; use cilium.io/v2 CiliumNodeConfig" subsys=klog
```

Fixes: db0d41af34 ("crd, option: Make CiliumNodeConfig available on v2 API")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

derailed reviewed (approved):
@christarazi Good catch!

christarazi committed:
option,resolver: Sort config sources descriptions in tests

This eases the subsequent commit when we update the test cases for
preferring the v2 version of CNC over v2alpha1.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

squeed commented on pkg/option/resolver/resolver.go:
```diff
@@ -89,6 +89,7 @@ func ResolveConfigurations(ctx context.Context, client client.Clientset, nodeNam
 		}
 	}
 
+	sort.Strings(sourceDescriptions)
```
I don't think this should be sorted; the ordering of sources matters and could be relevant while debugging.

christarazi replied:
This is sorted for unit testing reasons. Why does the order matter?

doniacld replied:
Why not sort them in the UT?

christarazi replied:
Good idea, updated to sort them in the unit tests instead.

squeed commented on pkg/option/resolver/resolver.go:
```diff
@@ -62,9 +62,30 @@ func ResolveConfigurations(ctx context.Context, client client.Clientset, nodeNam
+	// Find if there are any KindNodeConfig v2 sources first and prefer that
+	// over v2alpha1 if so.
+	sourcesV2Only := make([]ConfigSource, 0, len(sources))
+	sourcesExceptV2 := make([]ConfigSource, 0, len(sources))
```
I don't understand why we have to track this; the apiserver itself will only serve a single version, and it will upconvert any v2alpha1 resources to v2 transparently.

The only reason we have to have a fallback in the first place is that we might beat the operator to the CRD upgrade, and we really really really cannot skip an existing CiliumNodeConfig, whatever the scenario.

christarazi replied:
> we really really really cannot skip an existing CiliumNodeConfig, whatever the scenario.

If that's the case, then wouldn't this entire PR be invalid? If there's a nuance here, then I think I'm missing the full details.

dylandreimerink replied:
Can we resolve this conversation?

christarazi replied:
I think we can't until Casey gets back. He's out for now, but we'll wait until he's back.

squeed replied:
@christarazi I agree the fallback in `readNodeConfigsAllVersions` is needed, but I don't think we need to track sources by version. So AIUI these separate lists are unnecessary. Correct me if there's a case I'm missing.

christarazi replied:
@squeed So the reason for tracking sources by version is that the iteration order of sources matters, so we must put the v2 versions before the v2alpha1 versions. The iteration order matters because `readNodeConfigsAllVersions` assumes that the v2 versions are handled first, and _then_ falls back to v2alpha1.

squeed replied:
@christarazi it's not possible for a source to be one version and not the other; the apiserver automatically converts between the two versions for resources of the same name.

We just need to know globally whether or not the apiserver supports `v2` versus `v2alpha1`

christarazi replied:
@squeed That makes sense and it's possible that you can have two sources of different names and also different versions (v2 / v2alpha1), right? The code here addresses that case, as shown by the unit tests. If I remove this code, the unit tests fail. If we don't ensure that v2 is handled first then we will continue to get warning logs about deprecation because v2alpha1 will always be fetched as well. In other words, the last commit states that if v2 is found, v2alpha1 is ignored completely (to avoid warning logs). I assume this is ok because we have deprecated v2alpha1 so the user should have migrated already.

squeed replied:
> two sources of different names and also different versions (v2 / v2alpha1), right?

Nope, resources are unique for `(group, namespace, name)`. Version is only used for on-the-wire data format.

christarazi replied:
Given that then I think this PR is built on a false premise. What's not clear to me though is how we do we avoid the deprecation warning messages. My understanding is that because v2alpha1 is marked as deprecated, the K8s informer will automatically issue a warning log, so it's not directly in our control. The only option I see is to add an exception to the CI suite specifically for that warning msg here: https://github.com/cilium/cilium/blob/1293a478675dec3e7ff097dc81283287ed8992f4/test/helpers/cons.go#L270. What this means practically for users is that they will see this warning log if v2alpha1 resources exist and until they migrate over to v2. That sounds reasonable enough to me?

doniacld reviewed (approved):
lgtm! 🙏
